pennylane/gradients/parameter_shift.py

```diff
@@ -245,42 +261,39 @@ def processing_fn(results):
             results = [qml.math.squeeze(res) for res in results]
 
         grads = []
-        start = 1 if unshifted_coeffs and f0 is None else 0
+        start = 1 if at_least_one_unshifted and f0 is None else 0
         r0 = f0 or results[0]
 
-        for i, (s, f) in enumerate(zip(shapes, fns)):
+        for num_tapes, coeffs, fn, unshifted_coeff in gradient_data:
 
-            if s == 0:
+            if num_tapes == 0:
                 # parameter has zero gradient
-                g = qml.math.zeros_like(results[0])
-                grads.append(g)
+                grads.append(qml.math.zeros_like(results[0]))
                 continue
 
-            res = results[start : start + s]
-            start = start + s
+            res = results[start : start + num_tapes]
+            start = start + num_tapes
 
-            if f is not None:
-                res = f(res)
+            # individual post-processing of e.g. Hamiltonian grad tapes
+            if fn is not None:
+                res = fn(res)
 
             # compute the linear combination of results and coefficients
             res = qml.math.stack(res)
-            g = qml.math.tensordot(res, qml.math.convert_like(gradient_coeffs[i], res), [[0], [0]])
+            g = qml.math.tensordot(res, qml.math.convert_like(coeffs, res), [[0], [0]])
 
-            if unshifted_coeffs:
-                # add on the unshifted term
-                g = g + unshifted_coeffs[i] * r0
+            if unshifted_coeff is not None:
+                # add the unshifted term
+                g = g + unshifted_coeff * r0

```

Comment:
A general check was performed previously for whether *any* operation had an unshifted term, and then `unshifted_coeffs[i]` was used/attempted to use, which caused the `IndexError` in #2833